compiler/rustc_privacy/src/lib.rs

```diff
@@ -619,71 +648,78 @@ impl<'tcx> EmbargoVisitor<'tcx> {
 impl<'tcx> Visitor<'tcx> for EmbargoVisitor<'tcx> {
     type NestedFilter = nested_filter::All;
 
-    /// We want to visit items in the context of their containing
+    /// We want to visit items in the context of their containing  
     /// module and so forth, so supply a crate for doing a deep walk.
     fn nested_visit_map(&mut self) -> Self::Map {
         self.tcx.hir()
     }
 
     fn visit_item(&mut self, item: &'tcx hir::Item<'tcx>) {
-        let item_level = match item.kind {
+        let item_ev = match item.kind {
             hir::ItemKind::Impl { .. } => {
-                let impl_level = Option::<Level>::of_impl(
+                let impl_ev = Option::<EffectiveVisibility>::of_impl(
                     item.owner_id.def_id,
                     self.tcx,
                     &self.effective_visibilities,
                 );
-                self.update(item.owner_id.def_id, impl_level)
+
+                if let Some(impl_ev) = impl_ev {
+                    self.changed |= self.effective_visibilities.update(
+                        item.owner_id.def_id,
+                        None,
+                        || {
+                            ty::Visibility::Restricted(
+                                self.tcx.parent_module_from_def_id(item.owner_id.def_id),
+                            )
+                        },
+                        impl_ev,
+                        Level::Direct,
+                        self.tcx,
+                    );
+                }
+
+                impl_ev
             }
             _ => self.get(item.owner_id.def_id),
         };
-
         // Update levels of nested things.
         match item.kind {
             hir::ItemKind::Enum(ref def, _) => {
                 for variant in def.variants {
-                    let variant_level = self.update(variant.def_id, item_level);
+                    self.update(variant.def_id, item_ev, Level::Reachable);
+                    let variant_ev = self.get(variant.def_id);
                     if let Some(ctor_def_id) = variant.data.ctor_def_id() {
-                        self.update(ctor_def_id, item_level);
+                        self.update(ctor_def_id, variant_ev, Level::Reachable);
                     }
                     for field in variant.data.fields() {
-                        self.update(field.def_id, variant_level);
+                        self.update(field.def_id, variant_ev, Level::Reachable);
                     }
                 }
             }
             hir::ItemKind::Impl(ref impl_) => {
                 for impl_item_ref in impl_.items {
-                    if impl_.of_trait.is_some()
```

Comment:
I suppose this is a useless condition. Trait associated items inherit nominal visibility from trait. If trait is public the second condition will be fulfilled, otherwise impl level will be `None` and `update` will do nothing.

Reply:
Yeah, it seems useless now.